keps/sig-auth/4601-authorize-with-selectors/README.md

```diff
@@ -0,0 +1,875 @@
+    // Since requirements can only limit the request, it is safe to authorize as unlimited request if the requirements
+    // are not understood.
+	Requirements []metav1.LabelSelectorRequirement
+}
+
+type FieldSelectorRequirement struct {
+	// key is the label key that the selector applies to.
+	Key string `json:"key" protobuf:"bytes,1,opt,name=key"`
+	// operator represents a key's relationship to a set of values.
+	// Valid operators are In, NotIn, Exists, DoesNotExist
+	// The list of operators may grow in the future.
+	// Webhook authors are encouraged to ignore unrecognized operators and assume they don't limit the request.
+	// The semantics of "all requirements are AND'd will not change, so other requirements can continue to be enforced.
+	Operator LabelSelectorOperator `json:"operator" protobuf:"bytes,2,opt,name=operator,casttype=LabelSelectorOperator"`
+	// values is an array of string values. If the operator is In or NotIn,
+	// the values array must be non-empty. If the operator is Exists or DoesNotExist,
+	// the values array must be empty. This array is replaced during a strategic
+	// merge patch.
+	// +optional
+	// +listType=atomic
+	Values []string `json:"values,omitempty" protobuf:"bytes,3,rep,name=values"`
+}
+
+
+```
+
+Importantly, if old webhook authorizers do not honor these new field, they will assume the broadest possible access and fail closed.
+If old in-cluster authorization does not include field and label selectors, the kube-apiserver will assume the broadest possible access and fail closed.
+
+### Node Authorizer Changes
+
+The node authorizer will `node_authorizer.go` to check the `spec.nodeName` for List requests to ensure it matches the
+`nodeName` from its username (`system:node:<nodeName>`).
+Then we will remove the 
+```go
+		// TODO: restrict to pods scheduled on the bound node once field selectors are supported by list/watch authorization
+		rbacv1helpers.NewRule(Read...).Groups(legacyGroup).Resources("pods").RuleOrDie(),
+```
+from `bootstrappolicy/policy.go`.
+
+### CEL Authorizer Changes
+
+While admission isn't supported on List, Watch, or DeleteCollection, it is reasonable to expect that secondary authorization
+checks may desire to use those verbs and leverage the field and label selector capabilities.
+To support this we will two congruent options similar to
+```go
+	"fieldSelector": {
+		cel.MemberOverload("resourcecheck_fieldselector", []*cel.Type{ResourceCheckType, cel.StringType}, ResourceCheckType,
+			cel.BinaryBinding(resourceCheckName))},
+    }
+```
+This will allow usage like `authorizer.group('').resource('pods').fieldSelector('spec.nodeName=foo').check('list').allowed()`.
+The parsing will happen during the call to `allowed` where we track errors and have means of handling them already.
+The `cel/library` authorizer will force errors when `fieldSelector` or `labelSelector` are used for verbs that the kube-apiserver
```

Comment:
should a `fieldSelector` or `labelSelector` call that fails parsing error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

should use of `fieldSelector` or `labelSelector` in conjunction with a verb that doesn't currently honor selectors error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

Reply:
> should a fieldSelector or labelSelector call that fails parsing error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

This seems to make it very difficult to discovery VAP (or binding) bugs.  While I can see a case that leniency makes more successes, I'd try to make it easy to discover bugs and error.

> should use of fieldSelector or labelSelector in conjunction with a verb that doesn't currently honor selectors error, or should it drop the selector to be non-limiting (like the SubjectAccessReview API)?

Sure, updated.

Reply:
> While I can see a case that leniency makes more successes, I'd try to make it easy to discover bugs and error.

Hmm... if the selector was a hard-coded string, erroring seems useful ("your syntax has a bug").

If the selector is using a field from the object as a label, I guess I'd want an invalid selector due to weirdness from the object field to just skip the tighter scoping I might have gotten from the selector.

I'm fuzzy on whether it's possible to write a CEL expression defensively to tolerate an error... I don't think it is. How would you recommend someone scope with a selector if possible but fall back to unscoped if a valid selector wasn't possible to construct.

Reply:
also, it feels weird for the CEL API to error when SubjectAccessReview doesn't for effectively the same field (rawSelector)

Reply:
> How would you recommend someone scope with a selector if possible but fall back to unscoped if a valid selector wasn't possible to construct.

convincing when combined with user input.  updated.